src/rgw/rgw_admin.cc

```diff
@@ -5834,11 +5834,20 @@ int main(int argc, const char **argv)
         return -EINVAL;
       }
 
-      if (perm_policy_doc.empty()) {
+      if (perm_policy_doc.empty() && infile.empty()) {
         cerr << "permission policy document is empty" << std::endl;
         return -EINVAL;
       }
 
+      if (!infile.empty()) {
+        bufferlist bl;
+        int ret = read_input(infile, bl);
+        if (ret < 0) {
+          cerr << "ERROR: failed to read input policy document: " << cpp_strerror(-ret) << std::endl;
+          return -ret;
+        }
+        perm_policy_doc = bl.to_str();
+      }
```

Comment:
Cant the bufferlist bl be directly passed to IAM::Policy, instead of converting it to a string and then again back to a bufferlist?

Reply:
@pritha-srivastava Here is string needed: https://github.com/ceph/ceph/pull/37665/files#diff-f3b81fb31115c2616085efcfd6a8daed67a94d1041a03af5de48cc572978f547R5864
I think there is no difference which conversion run first. What do you prefer?

Reply:
I am sorry, I'm unable to understand what you are pointing at here. What I meant is:
bufferlist bl;
if (!infile.empty()) {
        int ret = read_input(infile, bl);
        if (ret < 0) {
          cerr << "ERROR: failed to read input policy document: " << cpp_strerror(-ret) << std::endl;
          return -ret;
        }
      } else {
        bl = bufferlist::static_from_string(perm_policy_doc);
      }

and then use bl in the try catch block?

Reply:
@pritha-srivastava In this line https://github.com/ceph/ceph/blob/master/src/rgw/rgw_admin.cc#L5855 `perm_policy_doc` is being used. If we used `infile` arg `perm_policy_doc` will be empty so this function will receive an empty `perm_policy_doc`. Am I wrong?

Reply:
There you can check if infile is not empty then you can get perm_policy_doc from bl like below:
if (!infile.empty()) {
 perm_policy_doc = bl.to_str();
}
role.set_perm_policy(policy_name, perm_policy_doc);

Reply:
Yes but we can do `perm_policy_doc = bl.to_str();` one time and prevent multi `if` clauses in code. Do you have any point for doing what you said?

Reply:
Yes you can do that. I just want you to avoid bufferlist bl = bufferlist::static_from_string(perm_policy_doc); for the infile case.

Reply:
@pritha-srivastava Well which one do you prefer? We can also change this function to be like this to prevent multi `if` clauses.
```cpp
role.set_perm_policy(policy_name, bl.to_str());
```

Reply:
Either way is ok with me here. You can choose what you prefer.

Reply:
@pritha-srivastava What do you think about the new change?